### DISCUSSION TITLE
ERROR: No loader is configured for ".node" files when working with argon2

### DISCUSSION THREAD

**Discussion by drishit96** [remix-run/remix#2468] [Q&A] [answered]
I am getting this error when I try to use argon2 in my project. Any ideas how to fix this?
```

> dev
> remix dev

(node:15555) ExperimentalWarning: stream/web is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Watching Remix app in development mode...

✘ [ERROR] No loader is configured for ".node" files: node_modules/argon2/lib/binding/napi-v3/argon2.node

    node_modules/argon2/argon2.js:12:12:
      12 │ } = require("./lib/binding/napi-v3/argon2.node");
         ╵             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Build failed with 1 error:
node_modules/argon2/argon2.js:12:12: ERROR: No loader is configured for ".node" files: node_modules/argon2/lib/binding/napi-v3/argon2.node
💿 Built in 664ms

```

**Comment by kiliman** [comment, marked as answer]
Remix handles files only for types listed in *modules.ts*. If you need to support a different type, you'll need to use `patch-package` and update this file. Then submit a PR for inclusion.

https://github.com/remix-run/remix/blob/main/packages/remix-dev/modules.ts

**Reply by nyoung697** [reply]
Do you have an example of how to use `patch-package` to add another file type?

**Reply by kiliman** [reply]
* Install `patch-package`.
* Edit `node_modules/@remix-run/dev/modules.d.ts` 
* Copy+paste an existing entry and set the correct extension. Try to keep them sorted alphabetically to ease merges.
* Run `npx patch-package @remix-run/dev` (this will create the patch in the patches folder)
* Update package.json and add a `"postinstall": "patch-package"` script (this ensures the patch will be installed automatically)
* Make sure to commit your patch in your repo

The patch should look something like this:
```diff
// patches/@remix-run+dev+1.4.1.patch
diff --git a/node_modules/@remix-run/dev/modules.d.ts b/node_modules/@remix-run/dev/modules.d.ts
index 5ccbf7d..8497d2d 100644
--- a/node_modules/@remix-run/dev/modules.d.ts
+++ b/node_modules/@remix-run/dev/modules.d.ts
@@ -49,6 +49,10 @@ declare module "*.mp4" {
     let asset: string;
     export default asset;
 }
+declare module "*.node" {
+    let asset: string;
+    export default asset;
+}
 declare module "*.ogg" {
     let asset: string;
     export default asset;
```

**Reply by nyoung697** [reply]
Thank you!

**Reply by MoSattler** [reply]
Hey @kiliman @nyoung697, did you get this to work? I have a similar problem with a different library, see [this](https://github.com/remix-run/remix/discussions/4455).
```
✘ [ERROR] No loader is configured for ".node" files: node_modules/fsevents/fsevents.node

    node_modules/fsevents/fsevents.js:13:23:
      13 │ const Native = require("./fsevents.node");
```

but I couldn't fix it with the above approach. Is that something that needs to be addressed in esbuild directly?

**Reply by kiliman** [reply]
Sorry. I've never used `.node` files before, so didn't realize they were a special type of binary module, so the instructions above were for generic files types and you wanted to treat them like .css or .jpg.

As far as I know, Remix doesn't support .node files. They would need to enable that in their esbuild configuration.

**Reply by thadk** [reply]
I believe the `compiler/utils/loader.ts` file may have also needed to be patched [along the lines of adding .zip](https://github.com/remix-run/remix/pull/3985/files#diff-a9488b95882e42c9908c223b503cf1a01ad8bf1e15a3a14711c9f770f6476de8). 

I was able to add `.py` file extension files loaded as a string into JS using this `patch-package` method: I'm a bit stunned that it worked and is relatively maintainable. 

The error about the "no loader" looks exactly like the one for `vite` which would have had a simpler workaround. It took a long time to troubleshoot: I had to realize vite is only used in vitest in Remix/Epic-Stack but instead remix's loader was the source of the error.

**Comment by vramana** [comment]
I have encountered the same issue. I am using remix with vite. So the solution in my case is add the following to vite config

```
  optimizeDeps: {
    exclude: ["@node-rs/argon2-darwin-arm64"] 
  },
```

**Reply by taggon** [reply]
> I have encountered the same issue. I am using remix with vite. So the solution in my case is add the following to vite config
> 
> ```
>   optimizeDeps: {
>     exclude: ["@node-rs/argon2-darwin-arm64"] 
>   },
> ```

This is the most simple solution for the problem right now. Note that you may need to exclude additional packages to work with other platforms.
